cpp/src/replace/replace.cu

```diff
@@ -13,97 +31,571 @@
+ * @param[out] output_valid Valid mask associated with output_data
+ * @param[out] output_valid_count #valid in output column
+ * @param[in] nrows # rows in `output_data`
+ * @param[in] values_to_replace_begin Device pointer to the beginning of the sequence
+ * of old values to be replaced
+ * @param[in] values_to_replace_end  Device pointer to the end of the sequence
+ * of old values to be replaced
+ * @param[in] d_replacement_values Device array with the new values
+ * @param[in] replacement_valid Valid mask associated with d_replacement_values
+ *
+ * @returns
+ */
+/* ----------------------------------------------------------------------------*/
+template<class T,
+    bool input_has_nulls, bool replacement_has_nulls>
+__global__
+void replace_kernel(cudf::column_device_view input,
+                    cudf::mutable_column_device_view output,
+                    cudf::size_type * __restrict__ output_valid_count,
+                    cudf::size_type nrows,
+                    cudf::column_device_view values_to_replace,
+                    cudf::column_device_view replacement)
+                    {
+  const T* __restrict__ input_data = input.data<T>();
+  cudf::bitmask_type const * __restrict__ input_valid = input.null_mask();
+  T * __restrict__ output_data = output.data<T>();
+  cudf::bitmask_type * __restrict__ output_valid = output.null_mask();
+  const T* __restrict__ values_to_replace_begin = values_to_replace.data<T>();
+  const T* __restrict__ values_to_replace_end = values_to_replace.data<T>()
+      + values_to_replace.size();
+  const T* __restrict__ d_replacement_values = replacement.data<T>();
+  cudf::bitmask_type const * __restrict__ replacement_valid = replacement.null_mask();
+
+  cudf::size_type i = blockIdx.x * blockDim.x + threadIdx.x;
+
+  uint32_t active_mask = 0xffffffff;
+  active_mask = __ballot_sync(active_mask, i < nrows);
+  auto const lane_id { threadIdx.x % cudf::experimental::detail::warp_size };
+  uint32_t valid_sum { 0 };
+
+  while (i < nrows) {
+    bool output_is_valid = true;
+    uint32_t bitmask = 0xffffffff;
+
+    if (input_has_nulls) {
+      bool const input_is_valid { cudf::bit_is_set(input_valid, i) };
+      output_is_valid = input_is_valid;
+
+      bitmask = __ballot_sync(active_mask, input_is_valid);
+
+      if (input_is_valid) {
+        thrust::tie(output_data[i], output_is_valid) =
+            get_new_value<T, replacement_has_nulls>(i, input_data,
+                                                    values_to_replace_begin,
+                                                    values_to_replace_end,
+                                                    d_replacement_values,
+                                                    replacement_valid);
       }
-      if (T{0.0} == e) {
-         return T{0.0};
+
+    } else {
+      thrust::tie(output_data[i], output_is_valid) =
+          get_new_value<T, replacement_has_nulls>(i, input_data,
```

Comment:
This call looks identical to the one above. Can it be factored out to only have to be written once?

Reply:
Yes, and done.